### PR TITLE
Refactor declaration provider to separate request

### PR DESCRIPTION
**Pull request by Hjaltesorgenfrei**

### CODE REVIEW

File: client/src/DecProvider.ts [hunk 1]
```diff
@@ -2,42 +2,38 @@ import { pathToFileURL } from "url";
 import { CancellationToken, Declaration, DeclarationProvider, Location, Position, ProviderResult, TextDocument, Uri } from "vscode";
 import { RequestType, VersionedTextDocumentIdentifier } from "vscode-languageclient";
 import Client from "./client";
-import { CompletionItemCoq, CompletionItemCoqRequest, CompletionItemCoqResponse } from "./protocol/types";
+import { CompletionItemCoq, CompletionItemCoqRequest, CompletionItemCoqResponse, DeclarationLocationCoqRequest, DeclarationLocationCoqResponse } from "./protocol/types";
+import * as fs from 'fs';
```
Comment (Duckapple) [inline client/src/DecProvider.ts]: I think we prefer explicit imports, though I do get that associating it with the namespace improves legibility somewhat.
```suggestion
import { existsSync } from 'fs';
```

File: client/src/DecProvider.ts [hunk 2]
```diff
@@ -2,42 +2,38 @@ import { pathToFileURL } from "url";
...
-        let items = await this.sendCompletionItemsRequest(document.uri, document.version, position);
-        let item = items.find((i) => i.label === word);
-        let pathItems = item?.path?.split('.');
-        let filename = item?.path?.split('.').at(-1) + ".v";
-        let path = this.coqlibPath + `/theories/` + pathItems?.slice(1, -1).concat([filename]).join("/");
-        return new Location(Uri.parse(path), new Position(0, 0));
+        let requestedDeclaration = document.getText(wordRange);
+        let response = await this.sendDeclarationLocationRequest(document.uri, document.version, position, requestedDeclaration);
+        let path = response.path.replace(/\.vo$/, ".v");
+        if (path && fs.existsSync(path)) {
```
Comment (Duckapple) [inline client/src/DecProvider.ts]:
```suggestion
        if (path && existsSync(path)) {
```

File: language-server/dm/documentManager.ml [hunk 3]
```diff
@@ -258,6 +258,12 @@ let get_lemmas st pos =
   | Some context -> 
     Some (ExecutionManager.get_lemmas context)
 
+let get_location st pos requestedDeclaration =
+  match get_context st pos with
+  | None -> None
+  | Some context -> 
+    Option.map (fun (path, range) -> path, None) (ExecutionManager.get_location context requestedDeclaration)
```
Comment (Duckapple) [inline language-server/dm/documentManager.ml:265]: You *could* use Option.map over pattern matching the context, but it doesn't really matter, this is probably more legible.

**Review by Duckapple — CHANGES_REQUESTED** [review]

File: language-server/dm/documentManager.ml (hunk 3 above)
Comment (Hjaltesorgenfrei) [reply language-server/dm/documentManager.ml:265]: I think 2 maps would be a bit much

File: language-server/dm/executionManager.mli [hunk 4]
```diff
@@ -28,6 +28,7 @@ val is_remotely_executed : state -> sentence_id -> bool
 val get_proofview : state -> sentence_id -> Proof.data option
 val get_context : state -> sentence_id -> (Environ.env * Evd.evar_map) option
 val get_lemmas : (Environ.env * Evd.evar_map) -> completion_item list
+val get_location : (Environ.env * Evd.evar_map) -> string -> (string * ((int * int) * (int * int)) option) option
```
Comment (Duckapple) [inline language-server/dm/executionManager.mli]: Here the position tuple is `(int * int) * (int * int)`, but in DocumentManager it is `Position.t * Position.t`. Should they use a common import?

File: language-server/dm/executionManager.ml [hunk 5]
```diff
@@ -435,6 +435,20 @@ let get_lemmas ((env : Environ.env), (sigma : Evd.evar_map)) =
...
 
+let get_location ((env : Environ.env), (sigma : Evd.evar_map)) requestedDeclaration =
+  let open Printer in
+  let open Loadpath in
+  let open Libnames in
+  let open Nametab in
+  let result = ref (None) in
+  let display ref kind env c =
+    let name = (Pp.string_of_ppcmds (pr_global ref)) in
+    if Option.is_empty result.contents && name = requestedDeclaration  then
```
Comment (Duckapple) [inline language-server/dm/executionManager.ml]: We are really nitpicking today lol
```suggestion
    if Option.is_empty result.contents && name = requestedDeclaration then
```

File: language-server/vscoqtop/lspManager.ml [hunk 6]
```diff
@@ -277,6 +277,21 @@ let coqtopStepForward ~id params : (string * Dm.DocumentManager.events) =
...
+  let uri = textDocument |> member "uri" |> to_string in
+  let loc = params |> member "position" |> parse_loc in
+  let requestedDeclaration = params |> member "requestedDeclaration" |> to_string in
+  let st = Hashtbl.find states uri in
+  match Dm.DocumentManager.get_location st loc requestedDeclaration with
+  | None -> ()
+  | Some (path, range) ->
+    let result = `Assoc [
+      "path", `String path;
+    ] in
```
Comment (Duckapple) [inline language-server/vscoqtop/lspManager.ml]: We will need to remember to add `range` here at some point (or just add it to the assoc-list based on whether it is Some or None).

**Review by Duckapple — APPROVED** [review]

File: language-server/vscoqtop/lspManager.ml (hunk 6 above)
Comment (Hjaltesorgenfrei) [reply language-server/vscoqtop/lspManager.ml]: yes

**Pull request merged**